crates/miden-objects/src/account/component/metadata/storage/schema.rs

```diff
@@ -0,0 +1,1202 @@
+            if slot_name.id() == AccountStorage::faucet_sysdata_slot().id() {
+                return Err(AccountComponentTemplateError::ReservedSlotName(slot_name));
+            }
+
+            if map.insert(slot_name.clone(), schema).is_some() {
+                return Err(AccountComponentTemplateError::DuplicateSlotName(slot_name));
+            }
+        }
+
+        Ok(Self { fields: map })
+    }
+
+    /// Returns an iterator over `(slot_name, schema)` pairs in slot-id order.
+    pub fn iter(&self) -> impl Iterator<Item = (&StorageSlotName, &StorageSlotSchema)> {
+        self.fields.iter()
+    }
+
+    /// Returns a reference to the underlying fields map.
+    pub fn fields(&self) -> &BTreeMap<StorageSlotName, StorageSlotSchema> {
+        &self.fields
+    }
+
+    /// Builds the initial [`StorageSlot`]s for this schema using the provided initialization data.
+    pub fn build_storage_slots(
+        &self,
+        init_storage_data: &InitStorageData,
+    ) -> Result<Vec<StorageSlot>, AccountComponentTemplateError> {
+        self.fields
+            .iter()
+            .map(|(slot_name, schema)| schema.try_build_storage_slot(slot_name, init_storage_data))
+            .collect()
+    }
+
+    /// Returns an iterator over init-value requirements for the entire schema.
+    pub fn init_value_requirements(&self) -> SchemaRequirementsIter<'_> {
+        Box::new(
+            self.fields
+                .iter()
+                .flat_map(|(slot_name, schema)| schema.init_value_requirements(slot_name)),
+        )
+    }
+
+    pub(crate) fn validate(&self) -> Result<(), AccountComponentTemplateError> {
+        for (slot_name, schema) in self.fields.iter() {
+            if slot_name.id() == AccountStorage::faucet_sysdata_slot().id() {
+                return Err(AccountComponentTemplateError::ReservedSlotName(slot_name.clone()));
+            }
+
+            schema.validate(slot_name)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Serializable for AccountStorageSchema {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u16(self.fields.len() as u16);
+        for (slot_name, schema) in self.fields.iter() {
+            target.write(slot_name);
+            target.write(schema);
+        }
+    }
+}
+
+impl Deserializable for AccountStorageSchema {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let num_entries = source.read_u16()? as usize;
+        let mut fields = BTreeMap::new();
+
+        for _ in 0..num_entries {
+            let slot_name = StorageSlotName::read_from(source)?;
+            let schema = StorageSlotSchema::read_from(source)?;
+
+            if fields.insert(slot_name.clone(), schema).is_some() {
+                return Err(DeserializationError::InvalidValue(format!(
+                    "duplicate slot name in storage schema: {slot_name}",
+                )));
+            }
+        }
+
+        Ok(Self { fields })
```

Comment:
Related to the above comment: do we not need to validate the deserialized schema?

Reply:
Yeah, validation was only being done in `AccountComponentMetadata::new()` instead of at the schema level. Changed this now.